windows/RNSVG/PatternView.cpp

```diff
@@ -55,77 +55,112 @@ void PatternView::UpdateProperties(IJSValueReader const &reader, bool forceUpdat
   SaveDefinition();
 
   if (auto const &root{SvgRoot()}) {
-    root.InvalidateCanvas();
+    root.Invalidate();
   }
 }
 
 void PatternView::UpdateBounds() {
   if (m_patternUnits == "objectBoundingBox") {
-    Rect rect{GetAdjustedRect(m_bounds)};
+    com_ptr<ID2D1ImageBrush> brush{get_self<D2DBrush>(m_brush)->Get().as<ID2D1ImageBrush>()};
+
+    D2D1_RECT_F rect{GetAdjustedRect(m_bounds)};
     CreateBrush(rect);
   }
 }
 
 void PatternView::CreateBrush() {
-  auto const &canvas{SvgRoot().Canvas()};
+  auto const root{SvgRoot()};
 
-  Rect elRect{GetAdjustedRect({0, 0, canvas.Size().Width, canvas.Size().Height})};
+  D2D1_RECT_F elRect{GetAdjustedRect({0, 0, static_cast<float>(root.ActualWidth()), static_cast<float>(root.ActualHeight())})};
   CreateBrush(elRect);
 }
 
-void PatternView::CreateBrush(Windows::Foundation::Rect const &rect) {
-  auto const &canvas{SvgRoot().Canvas()};
-  auto const &resourceCreator{canvas.try_as<ICanvasResourceCreator>()};
+void PatternView::CreateBrush(D2D1_RECT_F rect) {
+  auto const &root{SvgRoot()};
+
+  com_ptr<ID2D1Device> device{get_self<D2DDevice>(root.Device())->Get()};
 
-  if (auto const &cmdList{GetCommandList(rect)}) {
-    Brushes::CanvasImageBrush brush{resourceCreator, cmdList};
+  if (auto const &cmdList{GetCommandList(device.get(), rect)}) {
+    D2D1_IMAGE_BRUSH_PROPERTIES brushProperties{D2D1::ImageBrushProperties(rect)};
+    brushProperties.extendModeX = D2D1_EXTEND_MODE_WRAP;
+    brushProperties.extendModeY = D2D1_EXTEND_MODE_WRAP;
 
-    brush.SourceRectangle(rect);
-    brush.ExtendX(Microsoft::Graphics::Canvas::CanvasEdgeBehavior::Wrap);
-    brush.ExtendY(Microsoft::Graphics::Canvas::CanvasEdgeBehavior::Wrap);
+    com_ptr<ID2D1DeviceContext> deviceContext{get_self<D2DDeviceContext>(root.DeviceContext())->Get()};
 
-    cmdList.Close();
+    com_ptr<ID2D1ImageBrush> imageBrush;
+    check_hresult(deviceContext->CreateImageBrush(cmdList.get(), brushProperties, imageBrush.put()));
+
+    auto transform{D2D1::Matrix3x2F::Translation({rect.left, rect.top})};
+    if (m_transformSet) {
```

Comment:
if `m_transformSet` used to essentially treat `m_transform` as a nullable type? Is `std::optional` possible instead?

Reply:
using the identity matrix as default instead